File: java/test/org/rasdaman/rasnet/client/ServiceResponseHandlerTest.java
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */

package org.rasdaman.rasnet.client;

import com.google.protobuf.Message;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.rasdaman.rasnet.exception.UnsupportedMessageType;
import org.rasdaman.rasnet.message.Test.TestReply;
import org.rasdaman.rasnet.message.internal.Internal;
import org.rasdaman.rasnet.util.PeerMessage;
import org.rasdaman.rasnet.util.PeerStatus;
import org.rasdaman.rasnet.util.ZmqUtil;
import org.zeromq.ZMQ;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.rasdaman.rasnet.message.Communication.MessageType;
import static org.rasdaman.rasnet.message.Communication.MessageType.Types.ALIVE_PING;
import static org.rasdaman.rasnet.message.Communication.MessageType.Types.SERVICE_RESPONSE;
import static org.rasdaman.rasnet.message.Communication.ServiceCallStatus;

public class ServiceResponseHandlerTest {
    @Rule
    public ExpectedException thrown = ExpectedException.none();

    private ZMQ.Context context;
    private ZMQ.Socket server;
    private ZMQ.Socket client;
    private String address;
    private String peerId;
    private PeerStatus peerStatus;
    private String errorMessage;
    private ServiceResponseHandler responseHandler;
    private ConcurrentHashMap<String, Map.Entry<String, Message>> serviceRequests;
    private ConcurrentHashMap<String, Map.Entry<Boolean, byte[]>> serviceResponses;
    private TestReply reply;

    private ArrayList<byte[]> createErrorResponse() {
        ArrayList<byte[]> result = new ArrayList<>();
        MessageType type = MessageType.newBuilder().setType(SERVICE_RESPONSE).build();

        result.add(type.toByteArray());
        result.add(peerId.getBytes());

        ServiceCallStatus status = ServiceCallStatus.newBuilder().setSuccess(false).build();
        result.add(status.toByteArray());
        result.add(errorMessage.getBytes());

        return result;
    }

    private ArrayList<byte[]> createInvalidResponse() {
        ArrayList<byte[]> result = new ArrayList<>();
        MessageType type = MessageType.newBuilder().setType(ALIVE_PING).build();

        result.add(type.toByteArray());
        result.add(peerId.getBytes());

        ServiceCallStatus status = ServiceCallStatus.newBuilder().setSuccess(false).build();
        result.add(status.toByteArray());
        result.add(errorMessage.getBytes());

        return result;
    }


    private ArrayList<byte[]> createSuccessResponse() {
        ArrayList<byte[]> result = new ArrayList<>();
        MessageType type = MessageType.newBuilder().setType(SERVICE_RESPONSE).build();

        result.add(type.toByteArray());
        result.add(peerId.getBytes());

        ServiceCallStatus status = ServiceCallStatus.newBuilder().setSuccess(true).build();
        result.add(status.toByteArray());
        result.add(this.reply.toByteArray());

        return result;
    }

    @Before
    public void setUp() throws Exception {
        this.reply = TestReply.newBuilder().setData("data").build();
        this.address = "inproc://some-address";
        this.peerId = "peerId";
        this.errorMessage = "errorMessage";

        this.context = ZMQ.context(1);

        this.client = context.socket(ZMQ.DEALER);
        this.client.setIdentity(this.peerId.getBytes());
        this.client.connect(address);

        this.server = context.socket(ZMQ.ROUTER);
        this.server.bind(address);

        this.peerStatus = mock(PeerStatus.class);

        this.serviceRequests = new ConcurrentHashMap<String, Map.Entry<String, Message>>();
        this.serviceResponses = new ConcurrentHashMap<String, Map.Entry<Boolean, byte[]>>();
        this.responseHandler = new ServiceResponseHandler(server, peerStatus, serviceRequests, serviceResponses);
    }


    @After
    public void tearDown() throws Exception {
        this.client.disconnect(address);
        this.client.close();

        //this.server.unbind(address);
        this.server.close();

        this.context.close();
    }

    @Test
    public void testCanHandle() throws Exception {
        ArrayList<byte[]> message = this.createErrorResponse();
        assertTrue(this.responseHandler.canHandle(message));
    }

    @Test
    public void testCanHandleFail() throws Exception {
        ArrayList<byte[]> message = new ArrayList<>();
        assertFalse(this.responseHandler.canHandle(message));

        message = this.createInvalidResponse();
        assertFalse(this.responseHandler.canHandle(message));
    }

    @Test
    public void testHandleThrowException() throws Exception {
        thrown.expect(UnsupportedMessageType.class);

        ArrayList<byte[]> message = this.createInvalidResponse();
        this.responseHandler.handle(message);
    }

    @Test
    public void testHandleError() throws Exception {
        this.serviceRequests.put(peerId, new AbstractMap.SimpleEntry<String, Message>("method", null));

        ArrayList<byte[]> message = this.createErrorResponse();
        this.responseHandler.handle(message);

        assertTrue(this.serviceRequests.isEmpty());
        assertEquals(1, this.serviceResponses.size());

        Map.Entry<Boolean, byte[]> serviceResponse = this.serviceResponses.get(peerId);
        assertNotNull(serviceResponse);
        assertFalse(serviceResponse.getKey());
        assertEquals(errorMessage, new String(serviceResponse.getValue()));
    }

    @Test
    public void testHandle() throws Exception {
        Internal.ServiceRequestAvailable requestAvailable = Internal.ServiceRequestAvailable.getDefaultInstance();
        ZmqUtil.send(client, requestAvailable);

        PeerMessage peerMessage = new PeerMessage();
        ZmqUtil.receiveFromPeer(server, peerMessage);

        this.serviceRequests.put(peerId, new AbstractMap.SimpleEntry<String, Message>("method", null));

        ArrayList<byte[]> message = this.createSuccessResponse();
        this.responseHandler.handle(message);

        assertTrue(this.serviceRequests.isEmpty());
        assertEquals(1, this.serviceResponses.size());

        Map.Entry<Boolean, byte[]> serviceResponse = this.serviceResponses.get(peerId);
        assertNotNull(serviceResponse);
        assertTrue(serviceResponse.getKey());

        TestReply reply = TestReply.parseFrom(serviceResponse.getValue());

        assertEquals(this.reply.getData(), reply.getData());
    }

}

File: rasnet/test/client/serviceresponsehandlertest.cc
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */

#include "../../../common/src/unittest/gtest.h"
#include "../src/common/zmqutil.hh"
#include "../src/client/serviceresponsehandler.hh"
#include "../src/messages/internal.pb.h"

#include "../testutilities.hh"

namespace rasnet
{
namespace test
{
using std::map;
using std::string;
using std::make_pair;
using std::pair;
using boost::shared_ptr;

class ServiceResponseHandlerTest:public ::testing::Test
{

protected:
    ServiceResponseHandlerTest():
        server(ctx, ZMQ_ROUTER), client(ctx,ZMQ_DEALER)
    {
        this->address = "inproc://clientping";
        this->peerId = "peerId";
        this->errorMessage = "errorMessage";

        client.setsockopt(ZMQ_IDENTITY, this->peerId.c_str(),
                          strlen(this->peerId.c_str()));

        this->server.bind(address.c_str());
        this->client.connect(address.c_str());

        this->dummyReply.set_lifetime(1);
        this->dummyReply.set_retries(2);

        this->serviceRequests.reset(new map<string, pair<const google::protobuf::Message*, string> >());
        this->serviceResponses.reset(new map<string, pair<bool,shared_ptr<zmq::message_t> > >);
        this->serviceMutex.reset(new boost::mutex());

        this->handler.reset(new rasnet::ServiceResponseHandler(server, serviceRequests,serviceResponses, serviceMutex));
    }

    virtual ~ServiceResponseHandlerTest()
    {
        this->client.disconnect(address.c_str());
        this->client.close();

      //  this->server.unbind(address.c_str());
        this->server.close();
    }

    std::vector<boost::shared_ptr<zmq::message_t> > createErrorResponse()
    {
        std::vector<boost::shared_ptr<zmq::message_t> > result;
        MessageType type;
        type.set_type(MessageType::SERVICE_RESPONSE);

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(type)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::stringToZmq(peerId)));

        ServiceCallStatus status;
        status.set_success(false);
        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(status)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::stringToZmq(errorMessage)));

        return result;
    }

    std::vector<boost::shared_ptr<zmq::message_t> > createSuccessResponse()
    {
        std::vector<boost::shared_ptr<zmq::message_t> > result;
        MessageType type;
        type.set_type(MessageType::SERVICE_RESPONSE);

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(type)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::stringToZmq(peerId)));

        ServiceCallStatus status;
        status.set_success(true);
        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(status)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(this->dummyReply)));

        return result;
    }


    std::vector<boost::shared_ptr<zmq::message_t> > createInvalidResponse()
    {
        std::vector<boost::shared_ptr<zmq::message_t> > result;
        MessageType type;
        type.set_type(MessageType::ALIVE_PING);

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(type)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::stringToZmq(peerId)));

        ServiceCallStatus status;
        status.set_success(false);
        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::protoToZmq(status)));

        result.push_back(boost::shared_ptr<zmq::message_t>(TestUtilities::stringToZmq("error")));

        return result;
    }

    zmq::context_t ctx;
    std::string address;
    std::string peerId;
    std::string errorMessage;
    zmq::socket_t server;
    zmq::socket_t client;

    //Just for testing, if the class is not used, use something else
    ConnectReply dummyReply;

    boost::shared_ptr<std::map<std::string, std::pair<const google::protobuf::Message*, std::string> > > serviceRequests;
    boost::shared_ptr<std::map<std::string, std::pair<bool,boost::shared_ptr<zmq::message_t> > > > serviceResponses;
    boost::shared_ptr<boost::mutex> serviceMutex;

    boost::shared_ptr<rasnet::ServiceResponseHandler> handler;
};

TEST_F(ServiceResponseHandlerTest, canHandle)
{
    std::vector<boost::shared_ptr<zmq::message_t> > message = this->createErrorResponse();
    ASSERT_TRUE(this->handler->canHandle(message));
}

TEST_F(ServiceResponseHandlerTest, canHandleFail)
{
    std::vector<boost::shared_ptr<zmq::message_t> > message;
    ASSERT_FALSE(this->handler->canHandle(message));

    message = this->createInvalidResponse();
    ASSERT_FALSE(this->handler->canHandle(message));
}

TEST_F(ServiceResponseHandlerTest, handleThrowException)
{
    std::vector<boost::shared_ptr<zmq::message_t> > message = this->createInvalidResponse();;
    ASSERT_ANY_THROW(this->handler->handle(message));
}

TEST_F(ServiceResponseHandlerTest, handleError)
{
    this->serviceRequests->insert(make_pair(peerId, pair<const google::protobuf::Message*, string>(NULL,"method")));

    std::vector<boost::shared_ptr<zmq::message_t> > message = this->createErrorResponse();
    ASSERT_NO_THROW(this->handler->handle(message));

    ASSERT_TRUE(this->serviceRequests->empty());
    ASSERT_EQ(1, this->serviceResponses->size());

    std::map<std::string, std::pair<bool,boost::shared_ptr<zmq::message_t> > > ::iterator it;
    it=this->serviceResponses->find(peerId);
    ASSERT_TRUE(it!=this->serviceResponses->end());

    //The peerId must be the same
    ASSERT_EQ(peerId, it->first);

    //We must have an error
    ASSERT_FALSE(it->second.first);
    //Parse the error
    std::string errorResult = ZmqUtil::messageToString(*(it->second.second));
    ASSERT_EQ(errorResult, errorMessage);
}

TEST_F(ServiceResponseHandlerTest, handle)
{
    //Send the request available message from the main thread to the bridge socket
    internal::ServiceRequestAvailable reqAvailable;
    ZmqUtil::send(client, reqAvailable);
    std::string clientId;
    boost::shared_ptr<BaseMessage> baseMessage;
    ZmqUtil::receiveFromPeer(server, baseMessage, clientId);

    //Add the service request to the list of available requests
    this->serviceRequests->insert(make_pair(peerId, pair<const google::protobuf::Message*, string>(NULL,"method")));

    std::vector<boost::shared_ptr<zmq::message_t> > message = this->createSuccessResponse();
    ASSERT_NO_THROW(this->handler->handle(message));

    ASSERT_TRUE(this->serviceRequests->empty());
    ASSERT_EQ(1, this->serviceResponses->size());

    std::map<std::string, std::pair<bool,boost::shared_ptr<zmq::message_t> > > ::iterator it;
    it=this->serviceResponses->find(peerId);
    ASSERT_TRUE(it!=this->serviceResponses->end());

    //The peerId must be the same
    ASSERT_EQ(peerId, it->first);

    //We must have a success message
    ASSERT_TRUE(it->second.first);
    //Parse the response
    ConnectReply repl;
    boost::shared_ptr<zmq::message_t> r=it->second.second;
    repl.ParseFromArray(r->data(), r->size());

    ASSERT_EQ(repl.lifetime(), dummyReply.lifetime());
    ASSERT_EQ(repl.retries(), dummyReply.retries());

    //Check if the ServiceResponse was received through the bridge
    boost::shared_ptr<BaseMessage> serviceResponseAvailableEnv;
    ASSERT_TRUE(ZmqUtil::receive(client, serviceResponseAvailableEnv));
    internal::ServiceResponseAvailable responseAvailable;
    ASSERT_TRUE(responseAvailable.ParseFromString(serviceResponseAvailableEnv->data()));
}
}
}
